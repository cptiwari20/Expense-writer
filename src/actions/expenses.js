import uuid from 'uuid';

// add expense action Creator
export const addExpense = ({
  description= '',
  note= '',
  amount= 0,
  createdAt=0
} = {})=> {
  return {
    type: 'ADD_EXPENSE',
    payload: {
      description, amount, note, createdAt, id: uuid()
    }
  }
};
export const deleteExpense =(id) => {
  return {
    type: 'DELETE_EXPENSE',
    id
  }
};

export const editExpense = (id, updates)=> {
  return {
    type: 'EDIT_EXPENSE',
    payload: {
      id, updates
    }
  }
};