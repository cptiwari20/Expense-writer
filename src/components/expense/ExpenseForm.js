import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends Component {
  constructor(props){
    super(props);
    const { expense } = props
    this.state = {
      description: expense ? expense.description : '',
      amount: expense ? (expense.amount / 100).toString() : '',
      note: expense ? expense.note : '',
      createdAt: expense ? moment(expense.createdAt) : moment(),
      calenderFocused: false,
      error: ''
    }  
  }
   
  onDateChange = (createdAt) => (
    this.setState({ createdAt })
  )
  onFocusChange = ({ focused }) => {
    this.setState({ calenderFocused: focused})
  }
  onAmountChange = e => {
    const amount = e.target.value;
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
     return this.setState( { amount })
    }
  }

  handleSubmit = (e) => {
    e.preventDefault();
    if(!this.state.description || !this.state.amount){
      return this.setState({error: 'Please provide some description and amount!'})
    }
    this.setState({error: ''});
    let { description, createdAt, amount, note} = this.state
    this.props.onSubmit({
      description,
      createdAt: createdAt.valueOf(),
      amount: parseFloat(amount, 10) * 100,
      note
    });
  }
  render(){
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
      <form onSubmit={this.handleSubmit}>
        <input 
          type='text'
          placeholder='Description'
          value={this.state.description}
          onChange={e => this.setState({description: e.target.value})}  
        />
        <input 
          type='number'
          placeholder='Amount'
          value={this.state.amount}
          onChange={this.onAmountChange}  
        />
        <SingleDatePicker 
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea 
          placeholder='Note'
          value={this.state.note}
          onChange={e => this.setState({note: e.target.value})}  
        >
        </textarea>
        <button type='submit'>Submit</button>
      </form>
      </div>
    )
  }
}

export default ExpenseForm;