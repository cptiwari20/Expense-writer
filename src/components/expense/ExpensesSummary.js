import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import getVisibleExpenses from '../../selectors/visibleExpenses';
import getExpensesTotal from '../../selectors/getExpensesTotal';

export class ExpenseSummary extends Component {

  visibleExpenses = () => this.props.expenses.length;

  expensesTotal = () => 
  getExpensesTotal(this.props.expenses);

  render(){
    return (
      <div className='page-header'>
        <div className='content-container'>
            <h1 className='page-header__title'> 
              Showing <span>{this.visibleExpenses()}</span> expense(s), totalling <span>{this.expensesTotal() / 100}</span> Rupees.
            </h1>
            <div className="page-header__actions">
              <Link to='/add-expense' className='button'> 
                Create New Expense
              </Link>
            </div>
        </div>
      </div>
    )
  }
};

const mapStateToProps = ({expenses, filters}) => ( { expenses: getVisibleExpenses(expenses, filters) })

export default connect(mapStateToProps)(ExpenseSummary);