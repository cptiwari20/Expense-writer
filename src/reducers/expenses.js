export default ( state = [], action) =>{
  switch (action.type){
    case 'ADD_EXPENSE':
      return [ ...state, action.payload ];
    case 'EDIT_EXPENSE':
      return state.map(expense => {
        if(expense.id === action.payload.id){
          return {
            ...expense,
            ...action.payload.updates
          };
        }else{
          return expense;
        }
      })
    case 'DELETE_EXPENSE': 
      return state.filter(({id}) => id !== action.id);
    default:
      return state;
  }
};