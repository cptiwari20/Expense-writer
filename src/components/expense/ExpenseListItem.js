import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { deleteExpense } from '../../actions/expenses';
import moment from 'moment';

export const ExpenseListItem = ({dispatch, description, amount, note, createdAt, id}) => (
  <div>
    <Link to={`/edit/${id}`}>
     <h2>{description}</h2>
    </Link>
    <p>
    {amount / 100}
   /- Rupee(s) Only
    -
    <span>
    {moment(createdAt).format('Do MMM YYYY')}</span>
    </p>
    <p>{note}</p>
    
  </div>
);

export default connect()(ExpenseListItem);