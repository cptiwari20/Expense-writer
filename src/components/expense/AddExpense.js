import React, { Component } from 'react';
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../../actions/expenses';

class AddExpense extends Component{
  render(){
    return(
      <div>
        <h2>Add Your New Expense.</h2>
        <ExpenseForm addExpense={this.props.addExpense}/>
        </div>
    )
  }
}
export default connect(null, { addExpense })(AddExpense);