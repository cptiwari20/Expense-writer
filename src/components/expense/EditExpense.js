import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startDeleteExpense } from '../../actions/expenses';

export class EditExpense extends Component {
  onSubmit = (updates) => {
    this.props.startEditExpense(this.props.expense.id, updates);
    this.props.history.push('/');
  }
  onRemove = () => {
    this.props.startDeleteExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  }
  render(){
    return(
      <div>
        <h2>Edit your Expense</h2>
        <ExpenseForm 
            expense={this.props.expense}
            onSubmit = { this.onSubmit }  />
        <button onClick={this.onRemove}>Remove Expense</button>
      </div>
    )
  }
}
const mapStateToProps = ({expenses}, ownProps) => {
  return {
    expense: expenses.find(expense => expense.id === ownProps.match.params.id)
  }
}
export default connect(mapStateToProps, {startEditExpense, startDeleteExpense})(EditExpense);