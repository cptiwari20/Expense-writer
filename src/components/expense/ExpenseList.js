import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../../selectors/visibleExpenses';

const ExpenseList = (props) => (
  <div>
    <h3>Expenses</h3>
    <table border="1">
    <tbody>
    {props.expenses.map((expense, i) =>(
      <ExpenseListItem key={i} {...expense} />
    ))}
    </tbody>
    </table>
  </div>
);

const mapStateToProps = ({expenses, filters}) => {
  return {
    expenses: getVisibleExpenses(expenses, filters)
  }
}

export default connect(mapStateToProps)(ExpenseList);