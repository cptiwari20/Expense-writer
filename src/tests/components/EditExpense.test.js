import React from 'react';
import { shallow } from 'enzyme';
import { EditExpense }from '../../components/expense/EditExpense';
import ExpenseForm from '../../components/expense/ExpenseForm';
import expenses from '../fixtures/expenses';

let wrapper, editExpense, deleteExpense, history;
beforeEach(() => {
  editExpense = jest.fn();
  deleteExpense = jest.fn()
  history = { push: jest.fn() };

  wrapper = shallow(
    <EditExpense 
      editExpense={editExpense}
      d 
      deleteExpense={deleteExpense}
      history={history} 
      expense={expenses[1]}
    />
    )
});

describe('Edit Expense', () => {
  it('should show the Edit ExpensePage Correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle Edit Expense onSubmit', () => {
    wrapper.find(ExpenseForm).prop('onSubmit')(expenses[1]);

    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1])
  });

  it('should handle Delete Expense onRemove', () => {
    wrapper.find('button').simulate('click');

    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(deleteExpense).toHaveBeenLastCalledWith(expenses[1].id)
  })
})