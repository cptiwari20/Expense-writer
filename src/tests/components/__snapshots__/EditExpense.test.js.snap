// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Expense should show the Edit ExpensePage Correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit your Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 500,
          "createdAt": 0,
          "description": "Pen",
          "id": "2",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
